crates/worker/src/partition_processor_manager.rs

```diff
@@ -505,15 +515,48 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
             watchdog.run(),
         )?;
 
+        let mut logs_version_watcher = self.metadata.watch(MetadataKind::Logs);
+        let mut partition_table_version_watcher = self.metadata.watch(MetadataKind::PartitionTable);
+
         self.health_status.update(WorkerStatus::Ready);
         loop {
             tokio::select! {
                 Some(command) = self.rx.recv() => {
                     self.on_command(command);
                 }
-                Some(update_processors) = self.incoming_update_processors.next() => {
-                    if let Err(err) = self.on_control_processors(update_processors).await {
-                        warn!("failed processing control processors command: {err}");
+                Some(control_processors) = self.incoming_update_processors.next() => {
+                    self.on_control_processors(
+                        control_processors.into_body(),
+                        processor_mgr_effect_tx.clone())?;
+                }
+                _ = logs_version_watcher.changed(), if self.control_processors.is_some() => {
+                    // logs version has changed. and we have a control_processors message
+                    // waiting for processing. We can check now if logs version matches
+                    // and if we can apply this now.
+                    let control_processors = self.control_processors.take().unwrap();
```

Comment:
If `self.on_control_processors` expects the control processors value to be in `self.control_processors`, then we wouldn't have to move here and pass into the method. If the version is not up to date enough, then the only thing we do is to reassign to `self.control_processors`.